tensorflow/python/ops/math_grad.py

```diff
@@ -161,22 +161,26 @@ def _SumGrad(op, grad):
         return [array_ops.tile(grad, input_shape), None]
 
   input_shape = array_ops.shape(op.inputs[0])
-  # TODO(apassos) remove this once device placement for eager ops makes more
-  # sense.
-  with ops.colocate_with(input_shape):
-    output_shape_kept_dims = math_ops.reduced_shape(input_shape, op.inputs[1])
-    tile_scaling = _safe_shape_div(input_shape, output_shape_kept_dims)
-  grad = array_ops.reshape(grad, output_shape_kept_dims)
-  return [array_ops.tile(grad, tile_scaling), None]
+
+  if not op.get_attr("keep_dims"):
+    # TODO(apassos) remove this once device placement for eager ops makes more
+    # sense.
+    with ops.colocate_with(input_shape):
+      output_shape_kept_dims = math_ops.reduced_shape(input_shape, op.inputs[1])
+    grad = array_ops.reshape(grad, output_shape_kept_dims)
+  return [array_ops.broadcast_to(grad, input_shape), None]
```

Comment:
I love the replacement of tile (and its weird interface) with broadcast_to (and its saner interface). I tried to do this in the past and we had to roll it back because it was a performance regression.

Let's try again.

Reply:
if broadcast_to is slower than tile then probably broadcast_to kernel needs to be optimized, possibly by replacing it with tile's kernel. They are essentially doing the same thing anyway